crates/driver/src/domain/competition/solution/interaction.rs

```diff
@@ -53,17 +53,28 @@ impl Interaction {
                     liquidity::Kind::Swapr(pool) => pool.base.router.into(),
                     liquidity::Kind::ZeroEx(pool) => pool.zeroex.address().into(),
                 };
-                // As a gas optimization, we always approve the max amount possible. This
-                // minimizes the number of approvals necessary, and therefore
-                // minimizes the approval fees over time. This is a
-                // potential security issue, but we assume that the router contract for protocol
-                // indexed liquidity to be safe.
-                vec![eth::Allowance {
-                    token: interaction.input.token,
-                    spender: address,
-                    amount: eth::U256::max_value(),
-                }
-                .into()]
+                vec![
+                    // First unset any existing allowance in case the token (e.g. USDT)
```

Comment:
USDT doesn't have this issue since it doesn't update the allowance if it's set to `U256::max` ([code](https://etherscan.io/token/0xdac17f958d2ee523a2206206994597c13d831ec7#code#L181)). This is only needed for tokens that require resetting the allowance to 0 before changing it and that don't have this "don't update max allowance" optimisation which is very rare I believe.

Reply:
Solvers can encode approvals themselves so we can't assume that the allowance of USDT would always be set to `U256::MAX`, though.

Reply:
Sure, but imo there are only two reasonable strategies
1. Set it to what you need for this settlement (in which case it's 0 at the end of the settlement)
2. Set it to max

I'd be surprised if solvers were using any other values.

Reply:
Having any amount of uncertainty or an off-by-1 error on the approvals would already cause issues. TBH if this logic  becomes too involved I would likely just scrap this PR since this error obviously doesn't happen very often.

Reply:
Oh, also another strategy a sophisticated solver could implement is to round the approval amount up until it includes more 0 bytes to save on gas for the calldata.